File: app/config/jwt.py
# from bson import ObjectId
from passlib.context import CryptContext
from jose import jwt
from fastapi.security import OAuth2PasswordBearer, HTTPBearer
from fastapi import Depends, HTTPException, status
from app.config.prisma import prisma
from datetime import datetime, timedelta
from app.config.settings import settings

pwd_context = CryptContext(schemes=["bcrypt"])
security = HTTPBearer()
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/signin")


def hash_password(password: str):
    return pwd_context.hash(password, scheme="bcrypt")


def verify_password(password: str, hashed_password: str):
    return pwd_context.verify(password, hashed_password)


def create_token(user_id):
    to_encode = {"user_id": user_id}
    expire = datetime.utcnow() + timedelta(
        minutes=settings.ACCESS_TOKEN_EXPIRES_IN_MINUTES
    )
    to_encode.update({"exp": expire})
    jwt_token = jwt.encode(
        to_encode, settings.JWT_SECRET_KEY, algorithm=settings.JWT_ALGORITHM
    )
    return jwt_token


async def get_current_user(token: str = Depends(oauth2_scheme)):
    try:
        decoded = jwt.decode(
            token, key=settings.JWT_SECRET_KEY, algorithms=settings.JWT_ALGORITHM
        )
        user_id = decoded.get("user_id")
        user = await prisma.user.find_first(where={"id": int(user_id)})
        return user
    except Exception as error:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail={"success": False, "message": "Invalid authentication token."},
        )

File: lib/jwt.js
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import prisma from './prisma.js';

const ACCESS_TOKEN_EXPIRES_IN_MINUTES = process.env
  .ACCESS_TOKEN_EXPIRES_IN_MINUTES
  ? parseInt(process.env.ACCESS_TOKEN_EXPIRES_IN_MINUTES)
  : 60; // fallback to 60 minutes if not set

/**
 * Hashes a password using bcrypt.
 * @param {string} password
 * @returns {string} hashed password
 */
export function hashPassword(password) {
  return bcrypt.hashSync(password, 10);
}

/**
 * Verifies a password against a hash using bcrypt.
 * @param {string} password
 * @param {string} hashedPassword
 * @returns {boolean}
 */
export function verifyPassword(password, hashedPassword) {
  return bcrypt.compareSync(password, hashedPassword);
}

/**
 * Creates a JWT token for the given user ID.
 * @param {string|number} userId
 * @returns {string} JWT
 */
export function createToken(userId) {
  const payload = { userId };
  const expireSeconds = ACCESS_TOKEN_EXPIRES_IN_MINUTES * 60;
  const token = jwt.sign(payload, process.env.JWT_SECRET, {
    expiresIn: expireSeconds,
  });
  return token;
}

/**
 * Fetches the current user from a JWT token.
 * @param {string} token - Bearer token (without 'Bearer ' prefix)
 * @returns {Promise<object|null>} user object or null
 * @throws {Error} if token invalid or user not found
 */
export async function getCurrentUser(token) {
  try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    const userId = decoded.user_id;
    const user = await prisma.user.findFirst({ where: { id: Number(userId) } });
    if (!user) {
      throw new Error('User not found');
    }
    return user;
  } catch (error) {
    const err = new Error('Invalid authentication token.');
    err.status = 401;
    err.data = { success: false, message: 'Invalid authentication token.' };
    throw err;
  }
}
